Refactor to make it easier for future weather API changes.

bolivierjr opened the pull request:
Moved the check_user function to be a User model method and changed the tests to support it.
Making new classes to use dependency inversion, to take place of the helper functions, so we can switch weather APIs more easily.
Added a query_current_weather and query_location as client functions for abstraction on weather services.
Adding tests for the new classes and changing plugin.py to use the new interface.

bolivierjr committed:
Moved the check_user function to be a User model method and changed the tests to support it.

bolivierjr committed:
Making new classes to use dependency inversion, to take place of the helper functions,  so we can switch weather apis more easily.
Added a query_current_weather as a client function for abstraction on weather services.
Adding tests for the new classes and changing plugin.py to use the new interface.

tedski commented on models/users.py:
```diff
@@ -1,16 +1,11 @@
-    Model,
-    IntegerField,
-)
+from peewee import SqliteDatabase, CharField, DateTimeField, Model, IntegerField, DatabaseError
```
consider using isort to auto-sort imports

tedski commented on models/users.py:
```diff
@@ -43,6 +38,18 @@ def create_tables(cls) -> str:
+                raise DatabaseError("Users db and table not created yet.")
+            user = cls.get(cls.nick == nick)
+        except User.DoesNotExist:
+            user = None
```
Rather than returning `None` and having to deal with the `Union[User, None]` and `user.format if user else 1` throughout the code, you should implement the concept of an Anonymous User.  That could be a User instance with sane defaults that you are already implying later in the code anyway.

Use the `default` argument in the `User` fields to set those defaults.  Then you don't need to call `User.check_user` each time.  Your code can instantiate the `User` object using `User(msg=msg)` and your constructor can use that to call its own `check_user` logic to decide whether this is an Anonymous or Saved User and return a User instance either way.

tedski commented on utils/services.py:
```diff
@@ -0,0 +1,82 @@
+from cachetools import cached, TTLCache
+
+
+ttl_cache = TTLCache(maxsize=64, ttl=900)
```
Consider making this configurable.

tedski commented on utils/services.py:
```diff
@@ -0,0 +1,82 @@
+
+    def get_current(self, format: int = 1) -> str:
+        """
+        Returns back a formatted display of the weather to output.
```
nit: This is covered by the Returns section.  A better docstring would explain what it does.

e.g.
`Formats the weather data for display`

tedski commented on utils/weather.py:
```diff
@@ -0,0 +1,168 @@
+
+class WeatherAPI(ABC):
+    """
+    Weather api interface and required methods to implement new weather api classes.
```
nit: docstring relies on point in time ("new").  Abstract base class docstrings are best written with the future developer of a new weather API as the audience.

e.g.
```
Base class for weather APIs.  This class can be subclassed to implement a new API service for weather data.

Contains 2 abstract methods that must be implemented in the child class:

`find_current_weather()` which should return the weather data in format:

    .. code:: json
        {blah}
```
etc.

tedski reviewed (changes requested):
Fantastic work abstracting out the API.  Excellent test coverage.  Well done, this is a huge improvement.

bolivierjr committed:
Adding isort to the project makefile & pre-commit.
Ran isort on the project directory.

bolivierjr committed:
Changed the name of User.check_user method to get_user to be better descriptive.
Made an AnonymousUser class and fitted it in to return from the User.get_user method, if a user is not found with a default format number.
Made changes to plugin.py to use the improved get_user method in the weather and set_weather commands.
Made changes to some docstrings to be more descriptive.

bolivierjr committed:
Made the TTL_Cache configurable with environment variables.

bolivierjr commented:
Had to make a couple of fixups but here it is! :D

tedski commented on models/users.py:
```diff
@@ -43,10 +38,26 @@ def create_tables(cls) -> str:
             log.info("Created the users table")
             return "Created users table."
 
+    @classmethod
+    def get_user(cls, nick: str) -> Union["User", "AnonymousUser"]:
+        user: Union[User, AnonymousUser]
+        try:
+            if not isfile(db_path) or not cls.table_exists():
+                raise DatabaseError("Users db and table not created yet.")
+            user: User = cls.get(cls.nick == nick)
+        except User.DoesNotExist:
+            user: AnonymousUser = AnonymousUser()
+
+        return user
```
This should be a utility function outside of the `User` class since it may return a non-`User` object.

bolivierjr committed:
Made the get_user method a utility function outside of the User model since it returns an Anonymous user object if user is not found.
Changed the tests accordingly.

tedski commented on utils/users.py:
```diff
@@ -0,0 +1,45 @@
+    """
+    user: Union[User, AnonymousUser]
+    try:
+        if not isfile(db_path) or not User.table_exists():
+            raise DatabaseError("Users db and table not created yet.")
```
At this point, since this is a utility function, you probably don't need to worry about the db stuff here.  Instead, you can probably catch whatever exception is raised when User() is called without a db backend since you have the safety net of an anonymous user.

tedski reviewed (approved):
Really great work!  :shipit:
